### PR TITLE
Arming: Add opt in checks for mission items

### PR DESCRIPTION
**Pull request by WickedShell**
This adds opt in checks for required mission items. The overall target for this is to catch normally expected mission items and check that they are present in an effort to make sure that the user planned appropriately. (Rally points have the extra restriction of actually being close enough that you could actually select them in RTL).

If there is anything else anyone wants we can add it to the list, the only other one I am loosely planning on was adding `MAV_CMD_DO_GO_AROUND`. If people want we can add a bit for just "are there any items planned".

This also brings in a singleton for AP_Rally because it made life easier.

Final (and off topic notes): This would actually best be implemented as a list of `MAV_CMD` items on the SD card that we scan, but then the check wouldn't function on targets without an SD card, and you would need a check to make sure the file was actually present.

### CODE REVIEW

File: libraries/AP_Arming/AP_Arming.cpp [hunk 1]
```diff
@@ -493,6 +504,81 @@ bool AP_Arming::manual_transmitter_checks(bool report)
...
+        }
+        AP_Rally *rally = AP::rally();
+        if (rally == nullptr) {
+            check_failed(ARMING_CHECK_MISSION, report, "No rally library present");
+            #if CONFIG_HAL_BOARD == HAL_BOARD_SITL
+                AP_HAL::panic("Mission checks requested, but no rally was allocated");
+            #endif // CONFIG_HAL_BOARD == HAL_BOARD_SITL
+            return false;
+        }
+        for (int mask = 1; mask < MIS_ITEM_CHECK_MAX; mask <<= 1) {
```
Comment (peterbarker) [inline libraries/AP_Arming/AP_Arming.cpp]: This stuff would be a lot shorter if we factored https://github.com/ardupilot/ardupilot/blob/master/libraries/AP_Mission/AP_Mission.cpp#L1815 and used it in a `default:` (or stacked `case`s)

Comment (WickedShell) [reply libraries/AP_Arming/AP_Arming.cpp]: I'm not following, can you expand on that? As far as I can tell those are very different use cases.

Comment (peterbarker) [reply libraries/AP_Arming/AP_Arming.cpp]: See the pair of commits here: https://github.com/peterbarker/ardupilot/commits/wickedshell/arming-mission (untested, very rough!)

Comment (WickedShell) [reply libraries/AP_Arming/AP_Arming.cpp]: FWIW that's actually a larger build :D (252 bytes). With that being said I do see the appeal/makes it easy to add more individual checks in the future. Are we fine with the increased size vs clarity perspective here?

**Comment by magicrub** [comment]
Peter's method looks much easier to maintain

**Comment by peterbarker** [comment]
On Sun, 30 Dec 2018, WickedShell wrote:

> @peterbarker @tridge's variant (which is based around using a table and presented here) comes in 112 bytes smaller and is actually easier still to extend
> :)

*shrug*.  It shouldn't be smaller - that's a mystery.  Having duplicate 
strings for the one mission command should be a net loss in bytes.

Using a consistent way of getting the name means we'll be consistent in 
what we call the item when presenting to the user.

Still, since I'm not exactly rushing forward with an investigation into 
the build-size problem, I'm not really all that fussed.  We can always 
change it later.

**Pull request merged**